### PR TITLE
Replacing Hibernate Search 5 with Hibernate Search 6

### PR DESCRIPTION
**Pull request by tadgh**
This PR upgrades to Hibernate Search 6.0.0.CR2 for the purposes of removing various CVEs caused by old versions, along with our shaded jars. 

* Replace all out of date properties with new versions
* Replace all POM references to old versions
* Remove the shaded ES jars as HS6 supports 7.10.0
* Convert "Bridges" to RoutingBinder and RoutingBridge as per new API. 
* Modernize all indexing annotations
* Refactor all search queries. 
* Add backwards compatible Lucene50 codec
* Add missing CLI dependency. 
* Replace `MappingProvider` with `AnalysisConfigurer` as per new API. 
* Remove home-rolled `SearchBox` class, and replace with HS6 `GeoBoundingBox`
* Remove the `Suggest Keywords` operation from all Providers. 
* Remove lucene-highlighter 
* Remove Embedded Elasticsearch. That project is now defunct, and replaced with [TestContainers](https://www.testcontainers.org/)
* Remove the hapi-elasticsearch-6 shaded project. 
* Refactor the ElasticsearchSvcImpl to work with ES 7.10
* Replace usages of non-hibernate-orm `deleteBy` methods, as Hibernate Search literally ignores them for the purposes of indexing. 

**Outstanding issues:** 
* Currently, max-ngram-diff on elasticsearch is set on a per-index basis, controlled in an index setting. HS6 exposes no way currently to dynamically set this before the index is created. However there is an [open enhancement that will be delivered in 6.1](https://hibernate.atlassian.net/browse/HSEARCH-3934) which will resolve this issue. We may be able to fix it in the interim with a spring lifecycle hook in which we could quickly create an index template before Hibernate search 6 bootstrapping. In the meantime, you can circumvent this issue by applying the following template to your Elasticsearch Cluster before HAPI is upgraded: 

```json
{
	"index_patterns": [
		"termconcept-*",
		"resourcetable-*"
	],
	"settings": {
		"index": {
			"max_ngram_diff": 50
		}
	}
}
```
 
Breaking changes: 
* HAPI now relies on elasticsearch 7.10. If you have an install requiring elasticsearch, upgrade to 7.10
* HAPI now relies on lucene 8.7.0
* Anybody upgrading to this version will need to completely reindex their fulltext data due to large scale API changes and data field format changes between versions of hibernate search. 
* Suggest keywords operation is deleted.

### CODE REVIEW

File: hapi-fhir-cli/hapi-fhir-cli-api/pom.xml [hunk 1]
```diff
@@ -259,8 +259,14 @@
 			<artifactId>awaitility</artifactId>
 			<scope>test</scope>
 		</dependency>
+        <dependency>
+            <groupId>org.rauschig</groupId>
+            <artifactId>jarchivelib</artifactId>
+            <version>1.0.0</version>
```
Comment (jamesagnew) [inline hapi-fhir-cli/hapi-fhir-cli-api/pom.xml:265]: The version should be declared in the root pom file `<dependencyManagement>` section

File: hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java [hunk 2]
```diff
@@ -90,14 +98,13 @@ public Properties jpaProperties() {
...
-		extraProperties.put("hibernate.search.lucene_version", "LUCENE_CURRENT");
-		extraProperties.put("hibernate.search.default.worker.execution", "async");
 
+		extraProperties.put(BackendSettings.backendKey(LuceneBackendSettings.ANALYSIS_CONFIGURER), HapiLuceneAnalysisConfigurer.class.getName());
+		extraProperties.put(BackendSettings.backendKey(LuceneIndexSettings.DIRECTORY_TYPE), "local-filesystem");
+		extraProperties.put(BackendSettings.backendKey(LuceneIndexSettings.DIRECTORY_ROOT), "target/lucenefiles");
+		extraProperties.put(BackendSettings.backendKey(LuceneBackendSettings.LUCENE_VERSION), "LUCENE_CURRENT");
 		if (System.getProperty("lowmem") != null) {
-			extraProperties.put("hibernate.search.autoregister_listeners", "false");
+			extraProperties.put(HibernateOrmMapperSettings.ENABLED, "false");
```
Comment (jamesagnew) [inline hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java:107]: Ah nice they finally renamed this property... "autoregister_listeners" has to be the most unintuitive name imaginable for the "completely disable everything" flag.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/lastn/ElasticsearchSvcImpl.java [hunk 3]
```diff
@@ -721,7 +720,7 @@ public void close() throws IOException {
 	private IndexRequest createIndexRequest(String theIndexName, String theDocumentId, String theObservationDocument, String theDocumentType) {
 		IndexRequest request = new IndexRequest(theIndexName);
 		request.id(theDocumentId);
-		request.type(theDocumentType);
+//		request.type(theDocumentType);
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/lastn/ElasticsearchSvcImpl.java:723]: nitpick: drop commented line if it is no longer needed

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 4]
```diff
@@ -536,6 +535,8 @@ private void expandConcepts(IValueSetConceptAccumulator theAccumulator, TermValu
 			conceptViews = myTermValueSetConceptViewDao.findByTermValueSetId(theTermValueSet.getId(), displayValue);
 			wasFilteredResult = true;
 		} else {
+			// TODO GGG HS: I'm pretty sure we are overfetching here.  test says offset 3, count 4, but we are fetching index 3 -> 10 here, grabbing 7 concepts.
+			//Specifically this test testExpandInline_IncludePreExpandedValueSetByUri_FilterOnDisplay_LeftMatch_SelectRange
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:539]: When this is merged I'll have a look at this, it's probably my bug

File: pom.xml [hunk 5]
```diff
@@ -742,11 +742,11 @@
 		<jsr305_version>3.0.2</jsr305_version>
 		<junit_version>5.6.2</junit_version>
 		<flyway_version>6.5.4</flyway_version>
-		<!--<hibernate_version>5.2.10.Final</hibernate_version>-->
-		<hibernate_version>5.4.22.Final</hibernate_version>
+		<hibernate_version>5.4.24.Final</hibernate_version>
```
Comment (jamesagnew) [inline pom.xml:745]: Add these library version changes (except any that are only used in scope `test`) to here: https://github.com/jamesagnew/hapi-fhir/blob/master/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_3_0/changes.yaml

**Review by jamesagnew — APPROVED** [review]

File: hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java [hunk 6]
```diff
@@ -23,10 +23,16 @@
 import ca.uhn.fhir.jpa.api.config.DaoConfig;
 import ca.uhn.fhir.jpa.model.config.PartitionSettings;
 import ca.uhn.fhir.jpa.model.entity.ModelConfig;
-import ca.uhn.fhir.jpa.search.LuceneSearchMappingFactory;
+import ca.uhn.fhir.jpa.search.HapiLuceneAnalysisConfigurer;
```
Comment (jamesagnew) [inline hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java:26]: One general comment I'm adding here for lack of a better spot:

Before merging, can you try uploading a CodeSystem to an ES-backed server.. then wiping elasticsearch and marking all resources for reindexing, just to confirm that we reload elastic properly? That process will presumably go on to become the migration path assuming it works.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/ValueSetExpansionR4Test.java [hunk 7]
```diff
@@ -826,62 +660,31 @@ public void testExpandTermValueSetAndChildrenWithCountWithClientAssignedId() thr
 
 		myTermSvc.preExpandDeferredValueSetsToTerminologyTables();
 
-		ValueSetExpansionOptions options = new ValueSetExpansionOptions()
-			.setOffset(0)
-			.setCount(23);
-		ValueSet expandedValueSet = myTermSvc.expandValueSet(options, valueSet);
+//		ValueSetExpansionOptions options = new ValueSetExpansionOptions()
+//			.setOffset(0)
+//			.setCount(23);
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/term/ValueSetExpansionR4Test.java:665]: Commenting these out will use the default offset and count values for this test. Below, you've also commented out the related assertions.

Did you forget to uncomment these? The test name indicates count is relevant to the test.

File: hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java (hunk 2 above)
Comment (tadgh) [reply hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java:107]: Yeah threw me for a loop for sure

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 8]
```diff
@@ -885,68 +886,41 @@ private Boolean expandValueSetHandleIncludeOrExcludeUsingDatabase(IValueSetConce
...
-		 */
-		FullTextQuery jpaQuery = em.createFullTextQuery(luceneQuery, TermConcept.class);
-
 		/*
 		 * DM 2019-08-21 - Processing slows after any ValueSets with many codes explicitly identified. This might
 		 * be due to the dark arts that is memory management. Will monitor but not do anything about this right now.
 		 */
-		BooleanQuery.setMaxClauseCount(SearchBuilder.getMaximumPageSize());
+		//BooleanQuery.setMaxClauseCount(SearchBuilder.getMaximumPageSize());
+		//TODO GGG HS looks like we can't set max clause count, but it can be set server side.
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:922]: What's the impact of this change?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 9]
```diff
@@ -1030,149 +1049,236 @@ private void handleFilter(String theCodeSystemIdentifier, QueryBuilder theQb, Bo
...
-
-		if (theFilter.getOp() == ValueSet.FilterOperator.ISA) {
-			ourLog.debug(" * Filtering on codes with a parent of {}/{}/{}", code.getId(), code.getCode(), code.getDisplay());
-			theBool.must(theQb.keyword().onField("myParentPids").matching("" + code.getId()).createQuery());
-		} else {
-			throw new InvalidRequestException("Don't know how to handle op=" + theFilter.getOp() + " on property " + theFilter.getProperty());
-		}
+	private void addFilterLoincCopyright3rdParty(SearchPredicateFactory f, BooleanPredicateClausesStep<?> b) {
+		//TODO GGG HS These used to be Term term = new Term(TermConceptPropertyBinder.CONCEPT_FIELD_PROPERTY_PREFIX + "EXTERNAL_COPYRIGHT_NOTICE", ".*");, which was lucene-specific.
+		//TODO GGG HS ask diederik if this is equivalent.
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:1150]: This looks correct to me. There are related tests (or should be) so if they're happy, so am I.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 10]
```diff
@@ -967,22 +941,27 @@ private Boolean expandValueSetHandleIncludeOrExcludeUsingDatabase(IValueSetConce
...
-		int firstResult = jpaQuery.getFirstResult();
+		SearchQuery<TermConcept> termConceptsQuery = searchSession.search(TermConcept.class)
+			.where(f -> finishedQuery).toQuery();
+
+		System.out.println("About to query:" +  termConceptsQuery.queryString());
+		List<TermConcept> termConcepts = termConceptsQuery.fetchHits(theQueryIndex * maxResultsPerBatch, maxResultsPerBatch);
+
+
+		int resultsInBatch = termConcepts.size();
+		int firstResult = theQueryIndex * maxResultsPerBatch;// TODO GGG HS we lose the ability to check the index of the first result, so just best-guessing it here.
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:960]: I think this meets our needs. This is only used in debug logging. If there's something more meaningful that can be logged here, by all means add it. : )

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 11]
```diff
@@ -1030,149 +1049,236 @@ private void handleFilter(String theCodeSystemIdentifier, QueryBuilder theQb, Bo
...
+				handleFilterRegex(f, b, theFilter);
 				break;
 		}
 	}
 
-	private void isCodeSystemLoincOrThrowInvalidRequestException(String theSystemIdentifier, String theProperty) {
-		String systemUrl = getUrlFromIdentifier(theSystemIdentifier);
-		if (!isCodeSystemLoinc(systemUrl)) {
-			throw new InvalidRequestException("Invalid filter, property " + theProperty + " is LOINC-specific and cannot be used with system: " + systemUrl);
+	private void handleFilterRegex(SearchPredicateFactory theF, BooleanPredicateClausesStep<?> theB, ValueSet.ConceptSetFilterComponent theFilter) {
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:1081]: Nitpick! Meaningful names make the code more readable.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 12]
```diff
@@ -1188,37 +1294,55 @@ private void addLoincFilterAncestorIn(String theSystem, BooleanJunction<?> theBo
...
+		searchByParentPids(f, b, terms);
 	}
 
+	private void searchByParentPids(SearchPredicateFactory f, BooleanPredicateClausesStep<?> b, List<Term> theTerms) {
+		List<Long> parentPids = convertTermsToParentPids(theTerms);
+		b.must(f.bool(innerB -> {
+			parentPids.forEach(pid -> innerB.should(f.match().field(theTerms.get(0).field()).matching(pid)));
+		}));
+
+//		parentPids.stream()
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:1335]: Nitpick! If this isn't needed, can it be removed?

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java [hunk 13]
```diff
@@ -1030,149 +1049,236 @@ private void handleFilter(String theCodeSystemIdentifier, QueryBuilder theQb, Bo
...
+			}
+			if (!value.startsWith("^") && !value.startsWith(".*")) {
+				value = ".*" + value;
+			} else if (value.startsWith("^")) {
+				value = value.substring(1);
+			}
+
+			Term term = new Term(TermConceptPropertyBinder.CONCEPT_FIELD_PROPERTY_PREFIX + theFilter.getProperty(), value);
+			RegexpQuery query = new RegexpQuery(term);
+			//TODO GGG HS write the equivalent ES Query here.
```
Comment (dmuylwyk) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:1104]: Lots of TODOs in this file. Let me know if/when it needs another review. : )

**Review by dmuylwyk — COMMENTED** [review]
To the extent I understand Lucene and ES, these changes look great.

As we discussed, changelog entries are required.

We also need clear migration instructions (previously called out by @jamesagnew).

I focused my review on `BaseTermReadSvcImpl` and related tests. There are a lot of TODOs outstanding in that file so let me know if/when you'd like me to take another look.

This was a really interesting set of changes to review, and no doubt a frustrating set of changes to implement. Well done!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/elastic/ElasticsearchHibernatePropertiesBuilder.java [hunk 14]
```diff
@@ -56,36 +61,50 @@ public ElasticsearchHibernatePropertiesBuilder setPassword(String thePassword) {
...
+		theProperties.put(BackendSettings.backendKey(ElasticsearchIndexSettings.ANALYSIS_CONFIGURER), HapiElasticsearchAnalysisConfigurer.class.getName());
+
+		//theProperties.put("hibernate.search.default.elasticsearch.host", myRestUrl);
+		theProperties.put(BackendSettings.backendKey(ElasticsearchBackendSettings.HOSTS), myRestUrl);
+		theProperties.put(BackendSettings.backendKey(ElasticsearchBackendSettings.PROTOCOL), myProtocol);
 
-		theProperties.put("hibernate.search.default.elasticsearch.host", myRestUrl);
 		if (StringUtils.isNotBlank(myUsername)) {
-			theProperties.put("hibernate.search.default.elasticsearch.username", myUsername);
+			//theProperties.put("hibernate.search.default.elasticsearch.username", myUsername);
```
[the remark this answers is not in the document]
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/elastic/ElasticsearchHibernatePropertiesBuilder.java:75]: For sure, i just wanted eyes on the intial property swap. removing them all now.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java (hunk 8 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:922]: unclear to me at the moment, all tests continue to pass. Worth another look by somebody else.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java (hunk 11 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/BaseTermReadSvcImpl.java:1081]: These are super verbose and used all over the place, I think I am going to keep them as `f` and `b`, as this matches what is in the hibernate search docs, but i will add javadocs indicating what they are

File: hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java (hunk 6 above)
Comment (tadgh) [reply hapi-fhir-cli/hapi-fhir-cli-jpaserver/src/main/java/ca/uhn/fhir/jpa/demo/CommonConfig.java:26]: Done!

**Review by jamesagnew — APPROVED** [review]

**Comment by HananAwwad** [comment]
Hello @tadgh ,

Thanks alot for your pull request; really I started doing it on my local version; I did almost the same work you did here; again thanks a lot for your help; I will merge it in  the new version.
I just have a question in case I upgraded my ES; how could I re-index my current term concepts and resource tables which exists in the database to the new ES7 indexes. I used some thing called massIndexer from hibernate search 6 however I am wondering if mark-all-resources-for-reindexing operation will help me or its just related to search parameters.

**Comment by tadgh** [comment]
Yup, the `/$mark-all-resources-for-reindexing` endpoint will force everything to be reindexed with the new HS6 rules. It will work with all resources. I would love to see your PR that makes use of MassIndexer though, as that is probably a better way to go about this.

**Pull request merged**